[TensorExpr] Change API for creating Load and Store expressions.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45527 [WIP][TensorExpr] Add a tensor expressions tutorial.
* **#45520 [TensorExpr] Change API for creating Load and Store expressions.**

With this change `Load`s and `Store`s no longer accept `Placeholder`s in
their constructor and `::make` functions and can only be built with
`Buf`.
`Placeholder` gets its own `store`, `load`, `storeWithMask`, and
`loadWithMask` method for more convenient construction.

Differential Revision: [D23998789](https://our.internmc.facebook.com/intern/diff/D23998789)

Commit:
tutorial update.

[ghstack-poisoned]

Commit:
Update on "tutorial update."




[ghstack-poisoned]

Commit:
Update on "New work wihtout tutorial update"




[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Change API for creating Load and Store expressions."


With this change `Load`s and `Store`s no longer accept `Placeholder`s in
their constructor and `::make` functions and can only be built with
`Buf`.
`Placeholder` gets its own `store`, `load`, `storeWithMask`, and
`loadWithMask` method for more convenient construction.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Change API for creating Load and Store expressions."


With this change `Load`s and `Store`s no longer accept `Placeholder`s in
their constructor and `::make` functions and can only be built with
`Buf`.
`Placeholder` gets its own `store`, `load`, `storeWithMask`, and
`loadWithMask` method for more convenient construction.

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -1470,7 +1470,7 @@ void TensorExprKernel::bindInput(const torch::jit::Value* input) {
                   idx = idx + axes[i] * IntImm::make(*strides[i]);
                 }
-                return inBuffer(idx);
+                return inBuffer.load(idx);
```
Is `operator()` no longer supported for indexing into a buffer?

ZolotukhinM replied:
Correct, I removed it. It won't be as convenient anyway once we make `Placeholder`s mem-managed and they all become pointers. If we really need this functionality I think we might need to introduce somtheing like TensorHandle and PlaceholderHandle (it might looks ugly, but it'd be consistent with the rest of the system).

Commit:
Update on "[TensorExpr] Change API for creating Load and Store expressions."


With this change `Load`s and `Store`s no longer accept `Placeholder`s in
their constructor and `::make` functions and can only be built with
`Buf`.
`Placeholder` gets its own `store`, `load`, `storeWithMask`, and
`loadWithMask` method for more convenient construction.

Differential Revision: [D23998789](https://our.internmc.facebook.com/intern/diff/D23998789)

[ghstack-poisoned]

bertmaher commented on torch/csrc/jit/tensorexpr/tensor.h:
```diff
@@ -391,6 +387,21 @@ inline ExprHandle Tensor::call(const std::vector<T>& args) {
+inline ExprHandle Placeholder::load(const Ts&... ts) const {
+  std::vector<ExprHandle> params({ExprHandle(ts)...});
+  return ExprHandle(
+      new Load(data(), ExprHandleVectorToExprVector(params), new IntImm(1)));
```
Kind of OT and it's maybe inevitable but I always feel sad seeing these functions to convert-a-vector-of-one-thing-to-a-different-thing.  It feels like our APIs have internal impedance mismatches.  🤷

ZolotukhinM replied:
That's what we get for having handles and raw pointers :) Good thing is that's it's usually hidden in not-so-frequently changed functions.

nickgg reviewed (approved):
I'm happy with this if all the tests pass.

Commit:
Update on "[TensorExpr] Change API for creating Load and Store expressions."


With this change `Load`s and `Store`s no longer accept `Placeholder`s in
their constructor and `::make` functions and can only be built with
`Buf`.
`Placeholder` gets its own `store`, `load`, `storeWithMask`, and
`loadWithMask` method for more convenient construction.

Differential Revision: [D23998789](https://our.internmc.facebook.com/intern/diff/D23998789)

[ghstack-poisoned]
